ntp-udp/src/lib.rs

```diff
@@ -1,219 +1,424 @@
+            phantom: PhantomData<&'a [u8]>,
+        }
 
-pub(crate) fn zeroed_sockaddr_storage() -> libc::sockaddr_storage {
-    // a zeroed-out sockaddr storage is semantically valid, because a ss_family with value 0 is
-    // libc::AF_UNSPEC. Hence the rest of the data does not come with any constraints
-    unsafe { std::mem::MaybeUninit::zeroed().assume_init() }
-}
+        impl<'a> ControlMessageIterator<'a> {
+            // Safety assumptions:
+            // mhdr has a control and controllen field
+            // that together describe a memory region
+            // with lifetime 'a containing valid control
+            // messages
+            unsafe fn new(mhdr: libc::msghdr) -> Self {
+                // Safety:
+                // mhdr's control and controllen fields are valid and point
+                // to control messages.
+                let current_msg = unsafe { libc::CMSG_FIRSTHDR(&mhdr) };
+                Self {
+                    mhdr,
+                    current_msg,
+                    phantom: PhantomData,
+                }
+            }
+        }
 
-#[derive(Debug, Clone, Copy, Default)]
-pub(crate) struct TimestampingConfig {
-    rx_software: bool,
-    tx_software: bool,
-}
+        pub(crate) enum ControlMessage {
+            Timestamping(libc::timespec),
+            ReceiveError(libc::sock_extended_err),
+            Other(libc::cmsghdr),
+        }
 
-#[repr(C)]
-#[allow(non_camel_case_types)]
-#[derive(Default)]
-struct ethtool_ts_info {
-    cmd: u32,
-    so_timestamping: u32,
-    phc_index: u32,
-    tx_types: u32,
-    tx_reserved: [u32; 3],
-    rx_filters: u32,
-    rx_reserved: [u32; 3],
-}
+        impl<'a> Iterator for ControlMessageIterator<'a> {
+            type Item = ControlMessage;
+
+            fn next(&mut self) -> Option<Self::Item> {
+                // Safety:
+                // CMSG_FIRSTHDR and CMSG_NXTHDR only return valid pointers or NULL when given valid input
+                let current_msg = unsafe { self.current_msg.as_ref() };
+                if let Some(current_msg) = current_msg {
+                    // Safety:
+                    // New ensure mhdr is valid
+                    // CMSG_FIRSTHDR and CMSG_NXTHDR only return valid pointers or NULL when given valid input
+                    self.current_msg = unsafe { libc::CMSG_NXTHDR(&self.mhdr, self.current_msg) };
+
+                    match (current_msg.cmsg_level, current_msg.cmsg_type) {
+                        (libc::SOL_SOCKET, libc::SO_TIMESTAMPING) => {
+                            // Safety:
+                            // New ensures we have valid control messages
+                            // SO_TIMESTAMPING always has a timespec in the data
+                            let cmsg_data =
+                                unsafe { libc::CMSG_DATA(current_msg) } as *const libc::timespec;
+                            let timespec = unsafe { std::ptr::read_unaligned(cmsg_data) };
```

Comment:
Why/how do we know we need unaligned reads here?

Reply:
rather, we don't for sure know that the pointer is aligned. We could assume it, but why take the risk. `read_unaligned` will have the same performance as a `read` when the address does turn out to be aligned.

Reply:
Fair enough.